### PR TITLE
topology: sof-glk-es8336: initial support

### PR DESCRIPTION
**Pull request by plbossart**
Added from internal example

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]

**Comment by plbossart** [comment]
@hli25 can you try the TGL topology update? thanks!

**Comment by hli25** [comment]
> @hli25 can you try the TGL topology update? thanks!
yes, this TGL topology works too.
BTW, on TGL there is PCH-attached dmics.

**Comment by plbossart** [comment]
> > @hli25 can you try the TGL topology update? thanks!
> > yes, this TGL topology works too.
> > BTW, on TGL there is PCH-attached dmics.

@hli25 please try the DMIC update, that should be similar to what you tested?

**Comment by hli25** [comment]
> > > @hli25 can you try the TGL topology update? thanks!
> > > yes, this TGL topology works too.
> > > BTW, on TGL there is PCH-attached dmics.
> 
> @hli25 please try the DMIC update, that should be similar to what you tested?
Yes, DMIC works too on my TGL device with this topology.
btw, my GLK device work fine as before with this update.

**Comment by plbossart** [comment]
> > > > @hli25 can you try the TGL topology update? thanks!
> > > > yes, this TGL topology works too.
> > > > BTW, on TGL there is PCH-attached dmics.
> > 
> > 
> > @hli25 please try the DMIC update, that should be similar to what you tested?
> > Yes, DMIC works too on my TGL device with this topology.
> > btw, my GLK device work fine as before with this update.

great so now the only thing missing is HDMI/DP support. Do your devices have such connectors?

**Comment by hli25** [comment]
> > > > > @hli25 can you try the TGL topology update? thanks!
> > > > > yes, this TGL topology works too.
> > > > > BTW, on TGL there is PCH-attached dmics.
> > > 
> > > 
> > > @hli25 please try the DMIC update, that should be similar to what you tested?
> > > Yes, DMIC works too on my TGL device with this topology.
> > > btw, my GLK device work fine as before with this update.
> 
> great so now the only thing missing is HDMI/DP support. Do your devices have such connectors?

@plbossart 
The 2 devices in my hand has type-c port, so i can connect HDMI/DP monitor to them. Now a draft kernel patch was uploaded, most of the codes were borrowed from other existing machine drivers. But unfortunately, HDMI/DP has no sound output. [HDMI/DP support -- draft patch](https://github.com/hli25/linux-sof/commit/6ab1b3154c08a3b5d16412316d8663c2193948d7), and [logs](https://github.com/hli25/hli25/tree/pr-2962-HDMI-log). Please give some suggestion, thanks.

BTW, following is topology file update, without them there is no HDMI/DP device in the output of "aplay -l":

diff --git a/tools/topology/topology1/CMakeLists.txt b/tools/topology/topology1/CMakeLists.txt
index 12d9e6b5f..defc71fe9 100644
--- a/tools/topology/topology1/CMakeLists.txt
+++ b/tools/topology/topology1/CMakeLists.txt
@@ -156,8 +156,8 @@ set(TPLGS
        "sof-jsl-rt5682\;sof-jsl-rt5682-rt1015\;-DPLATFORM=jsl-rt1015"
        "sof-jsl-rt5682\;sof-jsl-rt5682-rt1015-xperi\;-DPLATFORM=jsl-rt1015\;-DINCLUDE_IIR_EQ=1"
        "sof-jsl-rt5682\;sof-jsl-rt5682-mx98360a\;-DPLATFORM=jsl-dedede"
-       "sof-glk-es8336\;sof-glk-es8336\;-DPLATFORM=bxt\;-DSSP_NUM=2\;-DCHANNELS=0\;-DHDMI=0"
-       "sof-glk-es8336\;sof-tgl-es8336\;-DPLATFORM=tgl\;-DSSP_NUM=0\;-DCHANNELS=2\;-DHDMI=0"
+       "sof-glk-es8336\;sof-glk-es8336\;-DPLATFORM=bxt\;-DSSP_NUM=2\;-DCHANNELS=0\;-DHDMI=1"
+       "sof-glk-es8336\;sof-tgl-es8336\;-DPLATFORM=tgl\;-DSSP_NUM=0\;-DCHANNELS=2\;-DHDMI=1"
 )
 
 add_custom_target(topologies1 ALL
diff --git a/tools/topology/topology1/sof-glk-es8336.m4 b/tools/topology/topology1/sof-glk-es8336.m4
index 4e6ff87e8..61bae0516 100644
--- a/tools/topology/topology1/sof-glk-es8336.m4
+++ b/tools/topology/topology1/sof-glk-es8336.m4
@@ -161,9 +161,9 @@ PCM_DUPLEX_ADD(ES8336, 0, PIPELINE_PCM_1, PIPELINE_PCM_2)
 
 ifelse(HDMI, `1',
 `
-PCM_PLAYBACK_ADD(HDMI 1, 5, PIPELINE_PCM_5)
-PCM_PLAYBACK_ADD(HDMI 2, 6, PIPELINE_PCM_6)
-PCM_PLAYBACK_ADD(HDMI 3, 7, PIPELINE_PCM_7)
+PCM_PLAYBACK_ADD(HDMI1, 5, PIPELINE_PCM_5)
+PCM_PLAYBACK_ADD(HDMI2, 6, PIPELINE_PCM_6)
+PCM_PLAYBACK_ADD(HDMI3, 7, PIPELINE_PCM_7)
 '
 )

**Comment by plbossart** [comment]
This PR is now ready for review thanks to the work of @hli25

Dependencies:
topology update in thesofproject/sof#4577
codec driver update in #3107 - this needs to be cleaned up by codec vendor and is not something the SOF team should review.

File: tools/topology/topology1/sof-glk-es8336.m4 [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+'
+)
+
+PCM_DUPLEX_ADD(ES8336, 0, PIPELINE_PCM_1, PIPELINE_PCM_2)
+
+ifelse(HDMI, `1',
+`
+PCM_PLAYBACK_ADD(HDMI 1, 5, PIPELINE_PCM_5)
+PCM_PLAYBACK_ADD(HDMI 2, 6, PIPELINE_PCM_6)
+PCM_PLAYBACK_ADD(HDMI 3, 7, PIPELINE_PCM_7)
```
Comment (kv2019i) [inline tools/topology/topology1/sof-glk-es8336.m4:166]: This won't work as HDMI is a variable, it gets replaced and you won't get "HDMI 1" in the output:

```
SectionPCM."1 1" {
        dai."1 1 5" {
                id "5"
        }
```

**Review by kv2019i — CHANGES_REQUESTED** [review]
M4 quoting issue in HDMI definition.

File: tools/topology/topology1/sof-glk-es8336.m4 (hunk 1 above)
Comment (plbossart) [reply tools/topology/topology1/sof-glk-es8336.m4:166]: thanks @kv2019i for finding this out.

**Comment by plbossart** [comment]
@hli25 please retest with HDMI no longer optional to make the m4 less complicated

**Review by kv2019i — APPROVED** [review]

**Comment by hli25** [comment]
> @hli25 please retest with HDMI no longer optional to make the m4 less complicated

HDMI/DP playback works with this topology, verified on my device.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@plbossart probably good to merge for v1.9 ? Just a conflict to sort..

**Comment by plbossart** [comment]
> @plbossart probably good to merge for v1.9 ? Just a conflict to sort..

@lgirdwood fixed conflict and fixed source of conflict: no reason why IMX/MTK topologies should be in the middle of Intel ones.

**Pull request merged**